.github/workflows/bin-ci.yml

```diff
@@ -0,0 +1,102 @@
+      with:
+        go-version: ${{ env.GO_VERSION }}
+    - name: golangci-lint
+      uses: golangci/golangci-lint-action@v3
+      with:
+        version: v1.45
+        args: |
+          --verbose
+        only-new-issues: false
+        skip-pkg-cache: true
+        skip-build-cache: true
+    - name: lint-consul-retry
+      shell: bash
+      run: |
+        go install github.com/hashicorp/lint-consul-retry@master && lint-consul-retry
+
+  test:
+    name: unit test (consul-version=${{ matrix.consul-version }})
+    strategy:
+      matrix:
+        consul-version:
+        - 1.12.0
+        - 1.12.0+ent
+    env:
+      TEST_RESULTS_DIR: /tmp/test-results
+      GOTESTSUM_VERSION: 1.8.0
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v3
+    - uses: actions/setup-go@v3
+      with:
+        go-version: ${{ env.GO_VERSION }}
+
+    - uses: actions/cache@v3
+      with:
+        path: |
+          ~/.cache/go-build
+          ~/go/pkg/mod
+        key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}
+        restore-keys: |
+          ${{ runner.os }}-go-
+
+    # This is broken I guess:
+    #   Error: Failed to fetch version metadata file HttpClientError: unsupported media type
+    #   requested, only [application/vnd+hashicorp.releases-api.v0+json
+    #   application/vnd+hashicorp.releases-api.v1+json] are available
+    #- name: Install Consul
+    #  uses: nickethier/action-setup-hashicorp-tool@main #TODO: update action when migrated to hc org
+    #  with:
+    #    product: consul
+    #    version: ${{ matrix.consul-version }}
+    - name: Install Consul
+      shell: bash
+      run: |
+        CONSUL_VERSION="${{ matrix.consul-version }}"
+        FILENAME="consul_${CONSUL_VERSION}_linux_amd64.zip"
+        curl -sSLO "https://releases.hashicorp.com/consul/${CONSUL_VERSION}/${FILENAME}" && \
+          unzip "${FILENAME}" -d /usr/local/bin && \
+          rm "${FILENAME}"
+        consul version
+    - name: Build
+      run: go build -v ./...
+    - name: Setup gotestsum
+      shell: bash
+      run: |
+        url=https://github.com/gotestyourself/gotestsum/releases/download
+        curl -sSL "${url}/v${{ env.GOTESTSUM_VERSION }}/gotestsum_${{ env.GOTESTSUM_VERSION }}_linux_amd64.tar.gz" | \
+        tar -xz --overwrite -C /usr/local/bin gotestsum
+    - name: Test
+      run: |
+        mkdir -p $TEST_RESULTS_DIR/${{ matrix.consul-version }}/json
+        PACKAGE_NAMES=$(go list ./... | grep -v 'mocks\|hack\|testing' | tr '\n' ' ')
+        echo "Testing $(echo $PACKAGE_NAMES | wc -w) packages"
+        if [[ "${{ matrix.consul-version }}" == *ent ]]; then
+          FLAG=-enterprise
+        fi
+        gotestsum \
+          --format=short-verbose \
+          --jsonfile $TEST_RESULTS_DIR/${{ matrix.consul-version }}/json/go-test-race.log \
+          --junitfile $TEST_RESULTS_DIR/${{ matrix.consul-version }}/gotestsum-report.xml \
+          -- $PACKAGE_NAMES -- $FLAG
```

Comment:
I disabled the race detector. We weren't running this before and it does cause test failures, which are largely due to races in the unit tests. I've fixed most of those, and will open a follow up to enable to race detector.